### ISSUE TITLE
Regarding the Agent Interface example.

### ISSUE THREAD

**Issue by lcipolina** [praveen-palanisamy/macad-gym#7]
Hello, in trying to run the example on the README.md for the Agent Interface, I did the following changes that you might want to consider:

#env = gym.make("HomoNComIndePOIntrxMASS3CTWN3-v0") # There is a typo here
env = gym.make("**HomoNcomIndePOIntrxMASS3CTWN3-v0**") #this is the correct one
#configs = env.configs() #this is a dict, cant assign to a variable

env_config = **env.configs**["env"]  #change this to read directly from dict
actor_configs = **env.configs[**"actors"] #change this to read directly from dict

I kept the rest of the code the same, but I got the following error after running exactly what is on the example, without changing anything else:

**Could not connect to Carla server because: rpc::rpc_error during call in function version**

**Comment by praveen-palanisamy** [comment]
Hi @lcipolina  ! Thank you for catching the typos and the error in the Agent Interface example. I have updated the README with the fixes.

After the fix, I copied the Agent Interface example into `example.py` and ran and it runs as expected. It will produce the outputs as shown below:
![macad-gym-example](https://user-images.githubusercontent.com/4770482/83512840-ac819300-a485-11ea-9c9e-55baca47c0a1.gif)

>Could not connect to Carla server because: rpc::rpc_error during call in function version

It looks like you may have forgotten to setup CARLA or set the `CARLA_SERVER` environment variable as mentioned in the [Getting Started section](https://github.com/praveen-palanisamy/macad-gym#getting-started).
If you already have CARLA binaries downloaded and extracted to your computer, please set the `CARLA_SERVER` env var. One way of doing this is shown below:

`echo "export CARLA_SERVER=${HOME}/software/CARLA_0.9.4/CarlaUE4.sh" >> ~/.bashrc`

**Comment by lcipolina** [comment]
Thank you very much for this. I was able to install CARLA's latest version, export the path to the "CarlaUE4.sh" and then running the test script from the command line.

I am using Teamviewer to connect to my lab's computer which has 40 GPUS (not SSH, but teamviewer)

I have encountered 2 problems, one I believe is from the remote connection (not sure) and the second one is some variable that I need to set up in Carla, maybe

<img width="663" alt="carla-errors" src="https://user-images.githubusercontent.com/5190890/83826709-2ee69e80-a6a2-11ea-9b4e-76ff160e7ff0.PNG">

**Comment by praveen-palanisamy** [comment]
Glad that you have got it running.

That's a pretty good machine you have got with 40 GPUs! :slightly_smiling_face:

Does the `macad-test.py` script that you are using hang after that?

**Comment by lcipolina** [comment]
actually no, I am able to use the WSAD commands, but I see no cars.

I guess the second error has to do with this:
https://github.com/carla-simulator/carla/issues/1942
https://carla.readthedocs.io/en/latest/adv_synchrony_timestep/

I will try tomorrow to fix it.

**Comment by praveen-palanisamy** [comment]
Okay then the scenario/training-env is not getting loaded. Which MACAD-Gym environment are you using?  
Is `macad-test.py` the same code as in the [Agent example](https://github.com/praveen-palanisamy/macad-gym#agent-interface)?

And yes, the warning message w.r.t the synchronous mode and `fixed_delta_seconds` is noted. It was added in one of the recent CARLA versions and I can push an update for taking care of that.

**Comment by praveen-palanisamy** [comment]
Also, could you try with [CARLA 0.9.4](https://drive.google.com/file/d/1p5qdXU4hVS2k5BOYSlEm7v7_ez3Et9bP/view)? Because this version was extensively tested and this version was the version which was used in the [NeurIPS 2019 Autonomous Driving Workshop paper](https://arxiv.org/abs/1911.04175) to get the results you were interested in. Once this works, we can fix the other issue to run with the latest CARLA version.

**Comment by lcipolina** [comment]
Hi yes, that's the code, the same as the Agent example.

The link for Carla0.9.4 posted here on GitHub is not downloading, not sure why. Maybe it's too big or something. I've tried several times. 

I can try to download it from the official site, but I am not sure which of these I need (I'm on Ubuntu)

[Linux] CARLA_0.9.9.2.tar.gz
[Linux] CARLA_0.9.9.2_RSS.tar.gz (RSS ready)

**Comment by praveen-palanisamy** [comment]
Alright.

[The link to CARLA 0.9.4 on this repo](https://drive.google.com/file/d/1p5qdXU4hVS2k5BOYSlEm7v7_ez3Et9bP/view) should be same as the one on CARLA's release page. Once you open this Google Drive link, there should be a download button. Did the download from the Google drive timeout or you ran into some other issue?
You can try downloading from [CARLA repo's release page for 0.9.4 here](https://github.com/carla-simulator/carla/releases/tag/0.9.4)

**Comment by lcipolina** [comment]
ok thank you I will try and report tomorrow. Many thanks!

**Comment by praveen-palanisamy** [comment]
Thank you for reporting the warning w.r.t synchronous mode with variable time step. I merged #8 with the update to use fixed time for stepping which should take care of that with recent versions of CARLA.

**Comment by lcipolina** [comment]
Hi, thanks for the amendment. I've downloaded the CARLA 0.94 file in a Windows machine using Chrome and it worked. I also did a git pull and the warning on the synchronous mode got resolved. 

Now I still face the issue with the TTY and the askpass (the first error on the red box) and the display is exactly the same. I don't see any car moving. 

_**sudo no tty present and no askpass program specified teamviewer**_

I am using Ubuntu 18 remotely using Teamviewer. I am searching online for the answer. Thank you very much!

**Comment by praveen-palanisamy** [comment]
Good to see the progress and glad that you have fixed the issue with sudo password entry.

Okay now the MACAD-Gym environment is loading! :tada: 

>This time the GUI opens and it moves (actually too much) when I move the cursor, to the point I can't see anything.

This is because, the CARLA UE4 window captures your mouse movements. You can prevent that by pressing the tilde `~` key when the CARLA window is active.

> `Could not connect to Carla server because: rpc::rpc_error during call in function version`

The above error is likely because there is a mismatch between your CARLA server version and the CARLA python client library version.
To solve it, you can do: `pip install carla==0.9.4` and then try running the script again.

**Comment by lcipolina** [comment]
Thank you very much Mr Palanisamy. I guess I have to uninstall my Carla 0.9.9 as the issue continues after the pip install 0.9.4?

**Comment by praveen-palanisamy** [comment]
If the same issue persists then yes, somehow there's still a version mismatch which likely means your client is using carla 0.9.9 while the server version is 0.9.4.

You can run this command: `python -m pip freeze | grep carla` to see which version of carla python client library version is active in your python environment.

**Comment by lcipolina** [comment]
Thank you very much. I was now able to start the environment and run the script. 
I have some questions if you don't mind. 

1. What is the action space of the agents?I am just guessing that the action space is the coordinates to move on? is speed included?

2. What is the reward of the agents?  I understand from the paper that this environment is a single-agent one and each maximizes it's own reward. Is the objective just to reach a certain point? are collisions penalized on the algo?

3. When running the sample script, without making any changes, I see that the cars collide at the intersection. I what happens

4. How can we plot the rewards as in your paper?
Thank you very much. 

<img width="320" alt="collision1" src="https://user-images.githubusercontent.com/5190890/83949281-72e3bb80-a7e8-11ea-859c-03a9a6cae263.PNG">

<img width="289" alt="collision2" src="https://user-images.githubusercontent.com/5190890/83949284-76774280-a7e8-11ea-9791-7e38434c145f.PNG">

<img width="276" alt="colission3" src="https://user-images.githubusercontent.com/5190890/83949288-7c6d2380-a7e8-11ea-980b-4231b6e00f37.PNG">

<img width="161" alt="collision4" src="https://user-images.githubusercontent.com/5190890/83949291-855df500-a7e8-11ea-9f2b-672458c8c897.PNG">

**Comment by praveen-palanisamy** [comment]
Nice to see that you have it working well on your machine!

> 1. What is the action space of the agents?I am just guessing that the action space is the coordinates to move on? is speed included?

You can check the action space of any MACAD-Gym environment using the `action_space` attribute like in an OpenAI Gym environment. The following is the minimal code:
```python
import gym
import macad_gym
env_name = "HomoNcomIndePOIntrxMASS3CTWN3-v0")
env = gym.make(env_name)
env.action_space
```
The action space for the `HomoNcomIndePOIntrxMASS3CTWN3-v0` will be:
`Dict(car1:Discrete(9), car2:Discrete(9), car3:Discrete(9))`

Which lists the action space for each agent as it is a Multi-agent environment.
For each individual car, the action corresponds to the following map:
```python
DISCRETE_ACTIONS = {
    # coast
    0: [0.0, 0.0],
    # turn left
    1: [0.0, -0.5],
    # turn right
    2: [0.0, 0.5],
    # forward
    3: [1.0, 0.0],
    # brake
    4: [-0.5, 0.0],
    # forward left
    5: [0.5, -0.05],
    # forward right
    6: [0.5, 0.05],
    # brake left
    7: [-0.5, -0.5],
    # brake right
    8: [-0.5, 0.5],
}
```

**Comment by praveen-palanisamy** [comment]
>What is the reward of the agents? I understand from the paper that this environment is a single-agent one and each maximizes it's own reward. Is the objective just to reach a certain point? are collisions penalized on the algo?

As described in the paper, this environment is a Multi-agent environment with independently acting agent. All the experiment details are in [Appendix C](https://arxiv.org/pdf/1911.04175.pdf#appendix.C) where the reward function is described as below:
![image](https://user-images.githubusercontent.com/4770482/83951225-cc49eb80-a7e4-11ea-80ae-edb1bb702df5.png)
In the `HomoNcomIndePOIntrxMASS3CTWN3-v0` environment the same reward function is used for each of the agent.

You can add your own reward function using the following as an example:
https://github.com/praveen-palanisamy/macad-gym/blob/5cb06a3278b049db8bdd8d768ed28d7fd929c11b/src/macad_gym/carla/reward.py#L21-L41

and also create your own environment if you would like to, based on [this wiki](https://github.com/praveen-palanisamy/macad-gym/wiki/creating-new-learning-environments.md).

**Comment by praveen-palanisamy** [comment]
> 3. When running the sample script, without making any changes, I see that the cars collide at the intersection. I what happens

That is because, in the sample agent script, for each of the car actors, the agent generate the action of driving forward/full-throttle.
This is the line where the environment is stepped with actions from the agents:
https://github.com/praveen-palanisamy/macad-gym/blob/d37741bfde03c5a9627a79f2572fcaa8a5742a43/examples/basic_agent.py#L38

And the `get_action()` method in the simple agent example returns the following action:
https://github.com/praveen-palanisamy/macad-gym/blob/d37741bfde03c5a9627a79f2572fcaa8a5742a43/examples/basic_agent.py#L25-L28

**Comment by praveen-palanisamy** [comment]
> 4. How can we plot the rewards as in your paper?

Oh, I missed this question. For the reward plots, I used [Tensorboard](https://www.tensorflow.org/tensorboard) using [this py package](https://github.com/lanpa/tensorboardX).

**Comment by lcipolina** [comment]
> 
> 
> > 1. How can we plot the rewards as in your paper?
> 
> Oh, I missed this question. For the reward plots, I used [Tensorboard](https://www.tensorflow.org/tensorboard) using [this py package](https://github.com/lanpa/tensorboardX).

Thank you very much for all these answers. So the rewards will be dumped in a file ? or should I produce the file output ?

**Comment by praveen-palanisamy** [comment]
You are welcome!
The sample agent script doesn't log the rewards to file. You will have to add the code to log the rewards to a file.

An example (feel free to use a different/simpler logger. This is just an example):
1. Import `tensorboardX` and create a summary writer:
```python
from tensorboardX import SummaryWriter
writer = SummaryWriter("logs")
```
2. Log the rewards
At every step you can log the reward to the log file using:
```python
writer.add_scalar("reward/actor1", total_reward_actor1)
```

3. Visualize the live reward plots
From a command line, run: `tensorboard --logdir=.` to visualize the log from the `logs` file.

**Comment by lcipolina** [comment]
Thank you very much for all this. My next task will be to reproduce the experiment on the paper, and then check whether I can run it on the latest Carla version. Thank you so much.

**Comment by praveen-palanisamy** [comment]
Closing this issue since the questions were answered. Please open a new issue for any further questions/issues. Thank you!
